diff --git a/libs/log/CMakeLists.txt b/libs/log/CMakeLists.txt
new file mode 100644
index 0000000..8659329
--- /dev/null
+++ b/libs/log/CMakeLists.txt
@@ -0,0 +1,28 @@
+#
+# Copyright Troy D. Straszheim
+#
+# Distributed under the Boost Software License, Version 1.0.
+# See http://www.boost.org/LICENSE_1_0.txt
+#
+#----------------------------------------------------------------------------
+# This file was automatically generated from the original CMakeLists.txt file
+# Add a variable to hold the headers for the library
+set (lib_headers 
+  log
+) 
+
+# Add a library target to the build system
+boost_library_project(
+  log
+  SRCDIRS src
+  #TESTDIRS  test
+  HEADERS ${lib_headers}
+  #  DOCDIRS 
+  DESCRIPTION "Lib for logging"
+  MODULARIZED 
+  #  AUTHORS 
+  #  MAINTAINERS 
+)
+
+
+
diff --git a/libs/log/module.cmake b/libs/log/module.cmake
new file mode 100644
index 0000000..87299ae
--- /dev/null
+++ b/libs/log/module.cmake
@@ -0,0 +1 @@
+boost_module(log DEPENDS date_time filesystem regex system thread)
diff --git a/libs/log/src/CMakeLists.txt b/libs/log/src/CMakeLists.txt
new file mode 100644
index 0000000..26170b0
--- /dev/null
+++ b/libs/log/src/CMakeLists.txt
@@ -0,0 +1,44 @@
+#
+# Distributed under the Boost Software License, Version 1.0.
+# See http://www.boost.org/LICENSE_1_0.txt
+#
+boost_add_library(log 
+        attribute_set.cpp
+        attribute_values_view.cpp
+        slim_string.cpp
+        code_conversion.cpp
+        core.cpp
+        stream_provider.cpp
+        severity_level.cpp
+        global_logger_storage.cpp
+        named_scope.cpp
+        process_id.cpp
+        timer.cpp
+        exceptions.cpp
+        sink_frontends.cpp
+        text_ostream_backend.cpp
+        text_file_backend.cpp
+        syslog_backend.cpp
+        thread_specific.cpp
+        trivial.cpp
+
+  DEPENDS boost_filesystem boost_thread boost_date_time
+  COMPILE_FLAGS "-DBOOST_LOG_USE_NATIVE_SYSLOG=1"
+  SHARED_COMPILE_FLAGS "-DBOOST_ALL_DYN_LINK=1"
+  NO_SINGLE_THREADED
+  )
+
+boost_add_library(log_setup 
+        parser_utils.cpp
+        init_from_stream.cpp
+        init_from_settings.cpp
+        filter_parser.cpp
+        formatter_parser.cpp
+
+  DEPENDS boost_filesystem boost_thread boost_date_time
+  COMPILE_FLAGS "-DBOOST_LOG_USE_NATIVE_SYSLOG=1"
+  SHARED_COMPILE_FLAGS "-DBOOST_ALL_DYN_LINK=1"
+  NO_SINGLE_THREADED
+  )
+
+
